src/TauBounds.cpp

```diff
@@ -19,9 +22,30 @@ void calculateTauBounds() {
 }
 
 void calculateTauLowerBounds() {
+  int mig, target, source;
+  double age;
+
+  for (int gen = 0; gen < dataSetup.numLoci; gen++) {
+    for (int i = 0; i < genetree_migs[gen].num_migs; i++) {
+      mig = genetree_migs[gen].living_mignodes[i];
+      target = genetree_migs[gen].mignodes[mig].target_pop;
+      source = genetree_migs[gen].mignodes[mig].source_pop;
+      age = genetree_migs[gen].mignodes[mig].age;
+      propagateLowerBound(target, age);
+      propagateLowerBound(source, age);
```

Comment:
this is quite inefficient, because you propagate for every migration event. It would be better to apply the same strategy you use for upper bounds. Set bounds directly for parents of source and target populations, and then propagate upward to ancestral populations only once after all loci have been processed.